cranelift-codegen/meta/src/isa/x86/registers.rs

```diff
@@ -24,7 +24,10 @@ pub(crate) fn define() -> IsaRegs {
     let builder = RegClassBuilder::new_toplevel("GPR", int_regs);
     let gpr = regs.add_class(builder);
 
-    let builder = RegClassBuilder::new_toplevel("FPR", float_regs);
+    let builder = RegClassBuilder::new_toplevel("FPR32", float_regs);
+    let fpr32 = regs.add_class(builder);
+
+    let builder = RegClassBuilder::subclass_of("FPR", fpr32, 0, 16);
```

Comment:
@bnjbvr, this (and the `.units(32)` above) is causing test failures that I don't understand:
 - [`isa::x86::registers::tests::unit_encodings`](https://github.com/bytecodealliance/cranelift/blob/21516e5e988e70d1ff453de7c0b22ad58fe464a0/cranelift-codegen/src/isa/x86/registers.rs#L27) is failing because now the GPRs miraculously have more units so `%xmm0` starts at `Some(32)`
 - similarly the [`unit_names`](https://github.com/bytecodealliance/cranelift/blob/21516e5e988e70d1ff453de7c0b22ad58fe464a0/cranelift-codegen/src/isa/x86/registers.rs#L47) test thinks the reg index `16` is invalid

What is going on?

Reply:
I think you're running afoul of https://github.com/bytecodealliance/cranelift/blob/5602d1564c4121e4703843f23cc083dfb29c5592/cranelift-codegen/meta/src/cdsl/regs.rs#L240-L247 which bumps the start of `FPR` to be aligned with its size. In theory you could add the expanded FPR first, then GPR, and everything should be okay?

Reply:
Aha!

Reply:
Hm, too quick to celebrate: I tried moving the `RegBankBuilder::new("FloatRegs"...` with the newly-expanded 32 units up above the `RegBankBuilder::new("IntRegs"...` but now the meta build crashes [due to r15](https://github.com/bytecodealliance/cranelift/blob/5602d1564c4121e4703843f23cc083dfb29c5592/cranelift-codegen/meta/src/isa/x86/recipes.rs#L430) with a backtrace of:
```
thread 'main' panicked at 'attempt to subtract with overflow', cranelift-codegen/meta/src/cdsl/regs.rs:60:25
stack backtrace:
...
  15: cranelift_codegen_meta::cdsl::regs::RegBank::unit_by_name
             at cranelift-codegen/meta/src/cdsl/regs.rs:60
  16: cranelift_codegen_meta::cdsl::regs::IsaRegs::regunit_by_name
             at cranelift-codegen/meta/src/cdsl/regs.rs:410
  17: cranelift_codegen_meta::isa::x86::recipes::define
             at cranelift-codegen/meta/src/isa/x86/recipes.rs:430
  18: cranelift_codegen_meta::isa::x86::define
             at cranelift-codegen/meta/src/isa/x86/mod.rs:61
  19: cranelift_codegen_meta::isa::define::{{closure}}
             at cranelift-codegen/meta/src/isa/mod.rs:62
...
```

Reply:
I think the failing assert is in error, f.ex `let reg_xmm1 = Register::new(fpr, regs.regunit_by_name(fpr, "xmm1"));` currently asserts for the same reason on master. It looks like `unit_by_name` shouldn't even trying to subtract from the register number (`15` in `r15`), and happens to work out so long as the specified reg class starts at 0 - subtract by 0 won't underflow.

I think instead `unit_by_name` ought to be checking just that `as_num < self.units`, and return `self.first_unit + as_num` like it would below?

Reply:
You're right. I pushed b0ee482 to fix that and b52fbdd to show what happens when I move the XMM registers first:
 - failures in `unit_names` and `unit_encodings` because, e.g., rax is no longer the first register; these can be fixed and are not a real problem
 - failures with unwinding (`test_*_alloc` and `windows_fastcall_x64_unwind.clif`): @peterhuene, it looks like the unwind info uses the register index directly; if I change a bunch of register indexes, does this completely break unwinding?
 - a failure in `simd-arithmetic-binemit.clif` is expected; that's what I'm currently working on
 - a failure in `pinned-reg.clif` like `panicked in worker #1: Couldn't clear fixed output interference for v1`; @bnjbvr, do you have any comment on this? Will changing the register indexes break regalloc?

Reply:
I was afraid there'd be some breakages like that :(

I noticed Windows unwinding using the regunit as an integer directly when fixing XMM preservation. Changing indices around will break it, but straightforward to fix: something like [this](https://github.com/bytecodealliance/cranelift/pull/1378/files#diff-c720e722fabb16edd5e9ce3554ee7f5dR85) but using `GPR.first`?

To the regalloc point, `pinned_reg` also accidentally depends on GPRs being first: [`is_pinned_reg`](https://github.com/bytecodealliance/cranelift/blob/master/cranelift-codegen/src/isa/registers.rs#L204-L209) tests against `pinned_reg` directly, where the ISA builder stores the provided `pinned_reg` [without translation](https://github.com/bytecodealliance/cranelift/blob/master/cranelift-codegen/meta/src/cdsl/regs.rs#L216-L220) and when building the x86 ISA description it's passed [the literal 15](https://github.com/bytecodealliance/cranelift/blob/master/cranelift-codegen/meta/src/isa/x86/registers.rs#L10). `pinned_reg` in the generated ISA should probably be offset from the start of its RegClass, so [`reg_bank.pinned_reg`](https://github.com/bytecodealliance/cranelift/blob/master/cranelift-codegen/meta/src/gen_registers.rs#L60) ought to be `reg_bank.pinned_reg + reg_bank.first_unit + reg_class.start`.

I think fixing dependence on `GPR.first == 0` would be a good PR independent of EVEX encoding, though I dunno how much of this will be replaced with New Backend Plans. (this is where stop talking and wait for Benjamin :D)

Reply:
@iximeow I think your proposed fix for the GPRs in the Windows unwind information makes sense.